keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+the kubeadm source tree.
+
+The current contents of the `k/k/cmd/kubeadm/app` directory
+should be moved in the root of k/kubeadm repository.
+
+The current `k/k/cmd/kubeadm/test` directory that contains
+integration tests should be merged with the `k/kubeadm/tests`
+directory under `k/kubeadm/tests/integration`.
+
+Existing tools in the k/kubeadm repository such as "kinder" and
+"operator" must continue existing as separate modules (sub-folders)
+inside the kubeadm module and can be obtained using
+`go get k8s.io/kubeadm/some-tool`.
+
+Existing docs and tests must co-exist in sub-folders too.
+
+Side effects:
+- All contents of k/kubeadm will be branched and tagged with kubeadm,
+but this is actually a positive change.
+- `hack/verify*` scripts for the root kubeadm folder should skip "kinder"
+and "operator". This area can see improvements in the future,
+potentially sharing scripts for all modules.
+
+#### Handling older kubeadm versions
+
+Currently kubeadm in the `release-1.17` branch (and later) of k/k
+does not depend on the rest of k/k. This was required for kubeadm
+to move out, however the older branches of k/k include versions of
+kubeadm that still depend on k/k.
+
+The proposed way forward is to move only the latest kubeadm from k/k master
+and build older version of kubeadm from k/k.
+
+This creates the following complications:
+- The post-submit job must handle tags and branches for kubeadm versions
+larger than X.
+- The kubernetes/release tooling must build kubeadm version X from the
+kubeadm repository while build older kubeadm from k/k.
+
+The cherry-pick script from k/k/hack will not work out of the box
+for the kubeadm repository so potentially it can be forked to suite
+kubeadm. The same cherry pick policy will apply as in k/k.
+
+#### Releasing and tagging
+
+##### Challenges
+
+The current plan is for kubeadm to continue to follow the k8s version scheme
+and release schedule. This will continue to make the skew of kubeadm
+and k8s easy to understand for the users and continue to give motivation
+to the kubeadm developers to get work done on time.
+
+This creates some complications:
+- Automation tools have to build kubeadm as part of the release
+- The k/kubeadm tags and branches need to follow those of k/k
+- k/k tags are not SEMVER, which can cause troubles when importing
+the k/kubeadm repository
+
+Currently kubeadm is released as part of the following k8s artifacts:
+- Node tarballs (Linux / Windows)
+- deb/rpm packages
+
+The kubeadm binary is built using either bazel (amd64 only) or using Docker
+(cross-build, all supported architectures).
+
+The binary is built with the same version as the rest of the released binaries
+like the kubelet. The version tooling resides in k/k under the /hack directory
+and it must be copied to k/kubeadm for consistency. The tooling that perform
+the packaging and building are location in k/release and require manual interaction.
+
+##### Building kubeadm from tarballs
```

Comment:
Could this be clarified as something like "End-user builds of kubeadm from tarballs"?